File: src/main/java/ml/a0x00000000/mjavascript/NodeJavaScript.java
package ml.a0x00000000.mjavascript;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.Log;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;

public class NodeJavaScript implements JavaScriptIsolation {
    public interface OnResultListener {
        void onResult(String text);
    }
    public interface OnReadyListener {
        void onReady();
    }
    public interface ExecutionListener {
        void onExecuted(@Nullable Exception err, @Nullable String res);
    }
    static public final String TAG = "NodeJavaScript";
    static public Thread nodeThread = null;
    static public int port = -1;
    static public boolean getReady() {
        return port != -1;
    }
    static private void setReady(boolean value) throws Exception {
        throw new Exception("Setter not available");
    }
    static private OnResultListener onResultCallback = null;

    // Used to load the 'native-lib' library on application startup.
    static {
        System.loadLibrary("native-lib");
        System.loadLibrary("node");
    }

    static private void copyAssets(Context context, String assetsName, String to) {
        InputStream inputStream = null;
        FileOutputStream fileOutputStream = null;
        try {
            String[] files = context.getAssets().list(assetsName);
            if(files.length > 0) {
                // Is a directory
                File file = new File(to);
                file.mkdirs();
                for (String filename : files) {
                    copyAssets(context, assetsName + "/" + filename, to + "/" + filename);
                }
            } else {
                // Is a file
                inputStream = context.getAssets().open(assetsName);
                fileOutputStream = new FileOutputStream(new File(to));
                byte[] buffer = new byte[1024];
                int count;
                while((count = inputStream.read(buffer)) != -1) {
                    fileOutputStream.write(buffer, 0, count);
                }
                fileOutputStream.flush();
                inputStream.close();
                fileOutputStream.close();
            }
        } catch (Exception err) {
            Log.e(TAG, "Ignored error: " + Log.getStackTraceString(err));
        } finally {
            if(null != inputStream) {
                try {
                    inputStream.close();
                } catch (Exception err) {
                    Log.e(TAG, "Ignored error: " + Log.getStackTraceString(err));
                }
            }
            if(null != fileOutputStream) {
                try {
                    fileOutputStream.close();
                } catch (Exception err) {
                    Log.e(TAG, "Ignored error: " + Log.getStackTraceString(err));
                }
            }
        }
    }

    public OnReadyListener onReadyCallback = null;
    private Context context = null;

    NodeJavaScript(Context context) {
        this.context = context;
        final String filesDir = context.getFilesDir().getAbsolutePath();
        copyAssets(context, "js", filesDir + "/js");
        if(null == nodeThread) {
            nodeThread = new Thread(new Runnable() {
                @Override
                public void run() {
                    int res = -1;
                    try {
                        res = startNodeWithArguments(new String[] {"node", filesDir + "/js/node-server.js" });
                    } catch(Exception err) {
                        Log.i(TAG, "Node.js was force closed");
                    }
                    Log.i(TAG, "Node.js exited with code " + res);
                    nodeThread = null;
                }
            });
            nodeThread.start();
            Log.i(TAG, "Node.js instance running");
        } else {
            Log.w(TAG, "Only one Node.js instance can be there at the same time");
        }
    }

    /**
     * Start a Node.js instance with arguments.
     * @param args Arguments to be passed to Node.js instance.
     * @return Exit status of Node.js instance.
     */
    native private int startNodeWithArguments(String[] args);

    @Override
    public void exec(@NotNull Execution _execution, @Nullable final Object callback) {
        NodeJavaScriptHelper.INSTANCE.exec(_execution, (ExecutionListener) callback);
    }

    @Override
    public void execFile(@NotNull String path, @org.jetbrains.annotations.Nullable Object callback) {
        // TODO: Not Implemented
    }

    @Override
    public void destroy() {
        if(null != nodeThread) {
            nodeThread.interrupt();
        }
        nodeThread = null;
    }

    /**
     * Handler function called by native code.
     * @param text String from Node.js script server.
     */
    static private void onResult(String text) {
        Log.v(TAG, "Incoming message: "+ text);
        if(null != onResultCallback) onResultCallback.onResult(text);
    }

    OnResultListener getOnResultCallback() {
        return onResultCallback;
    }

    /**
     * Set onResultCallback.
     * NOT THREAD-SAFE!
     * @param callback Callback function that handles returned result from Node.js server.
     */
    @Override
    public void setOnResultCallback(final Object callback) {
        if(null == onResultCallback && port == -1) {
            Log.i(TAG, "Setting shim callback");
            onResultCallback = new OnResultListener() {
                @Override
                public void onResult(String _port) {
                    Log.d(TAG, "In shim callback");
                    port = Integer.valueOf(_port.trim());
                    Log.i(TAG, "Target port: "+ port);
                    onResultCallback = (OnResultListener) callback;
                    Log.d(TAG, "onReadyCallback: " + onReadyCallback);
                    if(null != onReadyCallback) onReadyCallback.onReady();
                }
            };
        } else {
            onResultCallback = (OnResultListener) callback;
        }
    }
}

File: src/main/java/ml/a0x0000000000/mjavascript/NodeJavaScript.kt
package ml.a0x0000000000.mjavascript

import android.content.Context
import android.util.Log
import android.webkit.ValueCallback
import org.json.JSONObject
import java.io.*
import java.net.HttpURLConnection
import java.net.URL

class NodeJavaScript constructor(val context: Context): JavaScriptIsolation<ValueCallback<String>, ValueCallback<String>, (String) -> Unit> {
    companion object {
//        const val FILE_NAME: String = "NodeJavaScript"
        const val TAG: String = "NodeJavaScript"
        var nodeThread: Thread? = null
        var port: Int? = null

        // Used to load the 'native-lib' library on application startup.
        init {
            System.loadLibrary("native-lib")
            System.loadLibrary("node")
        }

        private fun copyAssets(context: Context, assetsName: String, to: String) {
            try {
                val files: Array<String> = context.assets.list(assetsName)
                if(files.isNotEmpty()) {
                    // Is a directory
                    val file = File(to)
                    file.mkdirs()
                    for(filename in files) {
                        copyAssets(context, "$assetsName/$filename", "$to/$filename")
                    }
                } else {
                    // Is a file
                    val inputStream = context.assets.open(assetsName)
                    val fileOutputStream = FileOutputStream(File(to))
                    val buffer = ByteArray(1024)
                    var count = inputStream.read(buffer)
                    while(count != -1) {
                        fileOutputStream.write(buffer, 0, count)
                        count = inputStream.read(buffer)
                    }
                    fileOutputStream.flush()
                    inputStream.close()
                    fileOutputStream.close()
                }
            } catch (err: Exception) {
                Log.e(TAG, Log.getStackTraceString(err))
            }
        }
    }

    private var _onMessageCallback: ((String) -> Unit)? = null
    var onMessageCallback: ((String) -> Unit)?
    get() {
        return _onMessageCallback
    }
    set(value) {
        if(null == _onMessageCallback && null == port) {
            _onMessageCallback = { _port -> port = _port.trim().toInt(); Log.i(TAG, "Target port: $port"); _onMessageCallback = value; onReadyCallback?.invoke() }
        } else {
            _onMessageCallback = value
        }
    }

    var onReadyCallback: (() -> Unit)? = null

    init {
        copyAssets(context, "js", "${context.filesDir}/js")
        if(null == nodeThread) {
            nodeThread = Thread(
                    Runnable {
                        var res = 0
                        try {
                            res = startNodeWithArguments(
                                    arrayOf(
                                            "node",
                                            "${context.filesDir}/js/node-server.js"
                                    )
                            )
                        } catch(err: InterruptedException) {
                            Log.i(TAG, "Node.js was force closed")
                        }
                        Log.i(TAG, "Node.js exited with code $res")
                        nodeThread = null
                    }
            )
            nodeThread!!.start()
            Log.i(TAG, "Node.js instance running")
        } else {
            Log.w(TAG, "Only one Node.js instance can be there at the same time")
        }
    }

    private fun onResult(text: String) {
        Log.i(TAG, "Incoming message: $text")
        onMessageCallback?.invoke(text)
    }

    /**
     * Start a Node.js instance with arguments.
     * @param args Arguments to be passed to Node.js instance.
     * @return Exit status of Node.js instance.
     */
    private external fun startNodeWithArguments(args: Array<String>): Int

    override fun exec(script: String, callback: ValueCallback<String>?) {
//        val file: File = createTempFile(FILE_NAME + System.currentTimeMillis(), ".js")
//        file.deleteOnExit()
//        file.writeText(script)
        val url = URL("http://127.0.0.1:$port")
        val connection: HttpURLConnection?
        val res: String
        val reader: BufferedReader
        val writer: BufferedWriter
        try {
            connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "POST"
            connection.doOutput = true
            connection.doInput = true
            connection.useCaches = false
            connection.setRequestProperty("Connection", "Keep-Alive")
            connection.readTimeout = 10000  // 10s
            connection.connectTimeout = 10000  // 10s
            connection.connect()

            val obj = JSONObject()
            obj.put("filename", "mJavaScript.js")
            obj.put("script", script)
            obj.put("workingDirectory", "${context.filesDir}")
            writer = connection.outputStream.bufferedWriter()
            writer.write(obj.toString())
            writer.close()

            reader = connection.inputStream.bufferedReader()
            res = reader.readText()
            Log.i(TAG, "Response: $res")
            reader.close()
            if(connection.responseCode != 200) {
                throw Exception(res)
            }
        } catch(err: Exception) {
            Log.e(TAG, "Cannot execute script: ${Log.getStackTraceString(err)}")
            throw Exception("Cannot execute script", err)
        }
    }

    override fun execFile(path: String, callback: ValueCallback<String>?) {
        TODO("NOT IMPLEMENTED")
    }

    override fun setOnResultCallback(callback: (String) -> Unit) {
        onMessageCallback = callback
    }

    override fun destroy() {
        nodeThread?.interrupt()
        nodeThread = null
    }
}